### ISSUE TITLE
Internal error regression in d77f0d9cf1bf breaking sail-riscv build

### ISSUE THREAD

**Issue by jrtc27** [rems-project/sail#156]
Compiling the attached code (note that GitHub has an extension whitelist for some stupid reason so I added a .txt to it...) using `sail -dmagic_hash -c -o test.out test.sail`, reduced from the full sail-riscv model (via -output_sail, hence the need for -dmagic_hash), gives the following error:

```
Internal error: Unreachable code (at "jib/jib_util.ml" line 672):
[test.sail]:260:23-35
260 |    input : bits(64) = rX_bits(rs1);
    |                       ^----------^
    | Invalid ctyp unifiers %bv32 and %bv64
    | 
    | Raised by primitive operation at Reporting.err_unreachable in file "reporting.ml", line 181, characters 18-43
    | Called from Reporting.unreachable in file "reporting.ml" (inlined), line 190, characters 8-35
    | Called from Jib_util.ctyp_unify in file "jib/jib_util.ml", line 672, characters 5-119
    | Called from Jib_compile.Make.compile_funcall.(fun) in file "jib/jib_compile.ml", line 638, characters 53-95
    | Called from Jib_compile.Make.compile_aexp in file "jib/jib_compile.ml", line 762, characters 26-62
    | Called from Jib_compile.Make.compile_aexp in file "jib/jib_compile.ml", line 964, characters 32-53
    | Called from Jib_compile.Make.compile_aexp.compile_case in file "jib/jib_compile.ml", line 798, characters 51-72
    | Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
    | Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
    | Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
    | Called from Jib_compile.Make.compile_aexp in file "jib/jib_compile.ml", line 817, characters 19-48
    | 
    | Please report this as an issue on GitHub at https://github.com/rems-project/sail/issues
```

According to a git bisect run this broke in d77f0d9cf1bf3612964eff5d8c290cc042233ef1, which seems like a rather likely culprit.

[test.sail.txt](https://github.com/rems-project/sail/files/7435300/test.sail.txt)

**Comment by Alasdair** [comment]
The always false constraint `assert(2 == 4, ...)` is confusing the backend, because for the flow-sensitive typing false implies everything, including 32 == 64, which allows the incorrect assignment on the line indicated by the internal error.

I have modified Sail so it can now detect whenever an assertion would make the typing context contain contradictory statements, and produce a warning. The typing consequences of these assertions are ignored, so the subsequent code produces a sensible error:

```
Warning: [mono_bug.sail]:259:4-57
259 |    assert(eq_int(2, 4), "repro-loop.sail 59:24 - 59:25");
    |    ^---------------------------------------------------^
    | 
Assertion would introduce contradictory (or outright false) premises into flow typing context.
This usually means the assertion is guaranteed to fail

Type error:
[mono_bug.sail]:260:23-35
260 |    input : bits(64) = rX_bits(rs1);
    |                       ^----------^
    | Tried performing type coercion from bitvector(32, dec) to bits(64) on rX_bits(rs1)
    | Coercion failed because:
    | Mismatched argument types in subtype check
    | 
```

However, I suspect this is treating the symptom of some other problem, rather than the root cause. Is the code in the example being generated by something?

**Comment by jrtc27** [comment]
`assert(sizeof(xlen) == 64)` and `assert(sizeof(xlen) >= 64)` are being used as the first statement in various places so we don't have to split out separate _rv64.sail versions of extension implementations (or, really, _not_rv32.sail; the `==` should probably also be `>=`) and can keep everything in the same file; some non-RV32 instructions are written in such a way that they assume there are 64 bits in GPRs.

**Comment by jrtc27** [comment]
(The decode clauses of course have xlen-based guards on them so unless you manually call the execute function for a specific instruction that code is genuinely unreachable)

**Comment by jrtc27** [comment]
So generating a warning on that pattern is definitely something that's _not_ wanted, as it's deliberately written to be an always-false assertion to squash typing errors and be a syntactic marker that there should be no code generated for that path beyond the assertion...

**Comment by Alasdair** [comment]
Ah ok, I see. In that case I suppose we want to do dead code elimination quite early on.

**Comment by jrtc27** [comment]
Yeah. I figured that’s what used to happen given it worked just fine, but maybe it was silently generating broken code that didn’t matter as it was still unreachable in the C output.

For simple tests it does seem like the dead code is eliminated early enough, I tried writing a reproducer from scratch but had no luck, hence the creduce’d mess provided here (which, curiously, got stuck in a 2-cycle when iteratively reducing).
